ocaml/typing/solver_intf.mli

```diff
@@ -118,21 +118,34 @@ module type Lattices_mono = sig
   val compose :
     'c obj -> ('b, 'c, 'd) morph -> ('a, 'b, 'd) morph -> ('a, 'c, 'd) morph
 
-  (* The following returns weaker than what we want, which is "\exists r.
-     allowed * r". But ocaml doesn't like existentials, and this weaker version
-     is good enough for us *)
-
-  (* Usual notion of adjunction f -| g :
-      [f x <= y] iff [x <= g y].
-
-     Our solver accepts a wider notion of adjunction and allows [f] and [g] to
-     be partial. They need to satisfy:
-     - If [f x <= y] and [y] is within [f]'s codomain, then [g] is well-defined
-     on [y] and [x <= g y].
-     - If not [f x <= y], we don't require anything about [g] on [y].
-     - If [y] is out of [f]'s codomain, we don't require anything about [g] on [y].
-
-     Similar in the other direction from [g] to [f]. *)
+  (* Usual notion of adjunction:
+     Given two morphisms [f : A -> B] and [g : B -> A], we require [f a <= b]
+      iff [a <= g b].
+
+     Our solver accepts a wider notion of adjunction and only requires the same
+     condition on convex sublattices. To be specific, if [f] and [g] form a
+      usual adjunction between [A] and [B], and [A] is a convex sublattice of
+     [A'], and [B] is a convex sublattice of [B'], we say that [f] and [g]
+     form a partial adjunction between [A'] and [B']. We do not require [f] to
+     be defined on [A'\A]; In fact, let's require [f] to be undefined on [A'\A]
+     for simplicity. Similar for [g].
```

Comment:
This still doesn't sound right to me. Could my sublattices be empty? Then any function is a partial adjunction to any other function. Or perhaps not empty: provided there exists an `a` in `A'` and `b` in `B'` such that `f a <= b` and `a <= g b`, then I think I can define `A = { a }` and `B = { b }` and declare that `f` and `g` are in partial adjunction. Maybe I'm missing something?

Reply:
What you said are correct. However, the sublattices being very small means the adjunction is less useful. In your second non-empty example, `f a` is a left mode only if `a` is from `A`. You are still allowed to apply `f` to `a'` where `a'` is from `A'` but not `A`, but the resulted mode `f a'` will not be a left mode, and not allowed to be on the LHS of submode. Now, this "now allowed" is not enforced by ocaml type checker, but kindly noted in the comment in `solver_intf.mli`.

Reply:
OK. I've made a note on internal ticket 2297.